### ISSUE TITLE
SoundSeer stops functioning when Mac awakes from sleep

### ISSUE THREAD

**Issue by podpah** [jonathangarelick/SoundSeer#37]
# Version
SoundSeer - v1.1.1
macOS - Sonoma 14.4.1
M1 Chip
# Issue
After the Mac goes to sleep, the icon stops updating and becomes an ear. The error shown is:
Not enough room for song. Try restarting
But when I restart it shows the song properly. This issue only occurs after the Mac has been sleeping
![image](https://github.com/jonathangarelick/SoundSeer/assets/107078260/14265b71-ebf7-48cb-82bb-312f673550f0)
It is fixed when you restart the app, but since it happens every time the Mac goes to sleep...

**Comment by jonathangarelick** [comment]
Hi there, thank you for using SoundSeer and for reporting this issue. Right now I am unable to replicate this. A couple questions:
- Are you using the version downloaded from the App Store?
- How much room is there between your menu bar icons and the physical "notch" ? It's hard to tell based on the image.

**Comment by podpah** [comment]
1 - I built from source
2 - I use it on a large monitor so there's no notch as my Mac stays closed (if I understand correctly and you're asking about the camera notch). I've tried with just the Mac and in that case the ear doesn't show up - if the right bar doesn't have enough space, SoundSeer isn't seen at all.

**Comment by jonathangarelick** [comment]
I see, thank you. In that case, it seems that there are too many menu bar icons. Unfortunately, Apple doesn't give me much leverage in this case--the operating system will kick icons off the menu bar and I am unable to control that. (If I remember correctly, it kicks off older ones first.)

Right now, I think the best I can do is what you're seeing--minimize the app to just a small icon to display a warning message. If you have any other suggestions though, I am happy to hear them.

**Comment by podpah** [comment]
I just did some testing around with it - it indeed is because there's too many menu bar icons for the laptop screen, even though I'm on just the display and it has more than enough space for it as it displays just fine with a ton of icons on display. 

On ideas - I'm not sure how it works but going off the error message and the fact it asks to restart because of the lack of space, maybe when it's on lack of space every 10/15s or so try and calculate the space again?

**Comment by jonathangarelick** [comment]
Oh, I see. You're opening SoundSeer when the laptop is the main display, then you switch over to your monitor?

It'll be a bit tricky, but what I can do is add a button to the menu to recalculate width. The issue is that I can only determine the max width when SoundSeer disappears, so if I run it at an interval, it will create a flashing animation when it disappears and appears again.

**Comment by podpah** [comment]
No, I never open the laptop. But since the problem is because there's no space on the laptop screen and it gets triggered on the display, I'm assuming when the Mac is waking up, the calculation gets done with the laptop display (even though it's not in use) before it loads on the display

**Comment by jonathangarelick** [comment]
Added to main: dc82f082f0dedd6e3e9e972b2f9d87949cf9fc5f

**Comment by podpah** [comment]
It partially solves it as it keeps displaying, but it doesn't go back to full size. So before sleep:
![image](https://github.com/jonathangarelick/SoundSeer/assets/107078260/1d18212c-8754-4543-8793-96f274d07cc6)
And after sleep:
![image](https://github.com/jonathangarelick/SoundSeer/assets/107078260/39277c1b-ef2c-4ad7-8646-7f4fd96f09b4)
But it doesn't revert back to full size afterwards
And if you're listening to really small title songs it also gets even smaller: (this was bingo • Don L)
![image](https://github.com/jonathangarelick/SoundSeer/assets/107078260/4ee05761-77b4-41e6-a694-d7e4e5a27a17)

**Comment by podpah** [comment]
The button worked once and then it stopped showing up even after restarting app/rebuilding (not sure if this is intended as I don't have experience with Swift Apps but it looks like the setting persisted) - it does partially work as a solution, but ends up being shortened on larger songs. Basing it on song length as it does normally might be a better setting than the fixed 45?
![image](https://github.com/jonathangarelick/SoundSeer/assets/107078260/ac8e8c0c-9be4-4d4e-b6c2-51d2d12e60dc)

**Comment by jonathangarelick** [comment]
Right, I should clarify this behavior. The "Reset Width" button will only appear when there's no room in the menu bar (i.e., when a song is playing and the ear icon is shown).

You are welcome to change the prefix length to something larger like 60, of course. I tried not to make it too large in order to be a good "neighbor" to other apps that have a menu bar presence.

**Comment by jonathangarelick** [comment]
Oh, I'm looking at your original comment in my emails. Yes, I can now see why it would be useful to permanently display an option to reset the width.

**Comment by podpah** [comment]
> Right, I should clarify this behavior. The "Reset Width" button will only appear when there's no room in the menu bar (i.e., when a song is playing and the ear icon is shown).

On this I meant that there was persistence with the setting through runs. After I clicked it the first time even after restarting the app/rebuilding it would act as the spotifyViewModel.prefixLength = 45 when I woke the Mac and reduce the size but not change to the ear (I might be going crazy)

And on this one:
> You are welcome to change the prefix length to something larger like 60, of course. I tried not to make it too large in order to be a good "neighbor" to other apps that have a menu bar presence.

I was under the impression that even though the length is set to something, if the title is smaller it'll automatically shrink? So when I was playing a song with a short title, when I clicked it, it seemed to shrink? I can get pictures on Monday to clarify what I mean, so I don't think making it large would be unneighbourly

**Comment by jonathangarelick** [comment]
I'm not entirely sure I understand your comment about persistence--how it works right now is that every time the app starts, the maximum text length is set to 45. If the app detects that it disappeared, it will gradually shrink itself until it either hits zero (ear icon) or appears again. Then it will never expand again until either (1) you use the button I posted, or (2) you restart the app.

I'm looking at your image ("I Smoked Away My Brain" song) and it's definitely not 45 characters long. What probably happened is that some external display weirdness caused it to shrink.

> if the title is smaller it'll automatically shrink?

As it's implemented now, it should only shrink when there's not enough room. `prefixLength` is not dependent on the song name itself.

You can make the "Reset Width" button permanently visible by taking it out of the if statement. Right now I am working through a refactor and I plan to permanently have a Reset Width button available just in case.

**Comment by jonathangarelick** [comment]
Oh, and I will also note that the app always removes everything after the parentheses in the song name.

So, `My Song (feat. Some Other Artist) - My Artist` becomes `My Song - My Artist`.
